gms/gossip_digest.hh

```diff
@@ -52,11 +53,11 @@ public:
         return _max_version;
     }
 
-    int compare_to(gossip_digest d) const {
+    std::weak_ordering compare_to(gossip_digest d) const {
```

Comment:
Is it really weak? Are there unequal digests that aren't greater or smaller than the other?

Reply:
Cc @michoecho 

https://en.cppreference.com/w/cpp/utility/compare/weak_ordering provides just what we need 

https://en.cppreference.com/w/cpp/utility/compare/strong_ordering provides a distinction between equal and equivalent that we don't need.

Reply:
> Is it really weak? Are there unequal digests that aren't greater or smaller than the other?

@avikivity Every strong ordering is also a weak ordering. It's the usage of "strong" that demands an additional justification, not the other way around.
By calling an ordering "strong", you assert that elements which compare equal are semantically the same element. Digests with the same `_generation` and `_max_version`, but with different `_endpoint` compare as equivalent. Do you think it's useful to communicate to the caller that they are the same digest?

Reply:
The name "weak ordering" is somewhat misleading. @michoecho is right in how he explains what weak/strong ordering is. The term that Avi was thinking about (that some items are not comparable) is called **partial** ordering (https://en.cppreference.com/w/cpp/utility/compare/partial_ordering), not weak ordering.

Reply:
We certainly should not compare gossip_digest without comparing the endpoint member. It's a recipe for later misunderstanding. For example, it will collapse items if stored in a std::map. If you want to compare these two members, extract them into a new struct. Or add the endpoint into the compare, whichever makes more sense.

I do think it's strong ordering. Two gossip_digest:s with the same generations are equal, not equivalent. The two objects are interchangable.

An example of equivalency is case insensitive compare or floating point compare. `A` and `a` are not the same thing, but they are equivalent when comparing case insensitively. +0 and -0 are equivalent, but not the same (floating point is in addition only partially ordered, because NaNs are unordered).

Reply:
@avikivity I agree about the need to compare the endpoint.
I think that the idea was that the caller checks the endpoint separately and then calls compare_to
although this is not documented anywhere.
Luckily, I realized that those compare_to functions are unused, so I'll just drop them.